velox/functions/sparksql/specialforms/FromJson.cpp

```diff
@@ -0,0 +1,602 @@
+        }
+        break;
+      }
+      default:
+        return simdjson::INCORRECT_TYPE;
+    }
+    return simdjson::SUCCESS;
+  }
+
+  template <typename To, typename From>
+  static simdjson::error_code convertIfInRange(
+      From x,
+      exec::GenericWriter& writer) {
+    static_assert(std::is_signed_v<From> && std::is_signed_v<To>);
+    if constexpr (sizeof(To) < sizeof(From)) {
+      constexpr From kMin = std::numeric_limits<To>::lowest();
+      constexpr From kMax = std::numeric_limits<To>::max();
+      if (!(kMin <= x && x <= kMax)) {
+        return simdjson::NUMBER_OUT_OF_RANGE;
+      }
+    }
+    writer.castTo<To>() = x;
+    return simdjson::SUCCESS;
+  }
+};
+
+template <TypeKind kind>
+simdjson::error_code parseJsonOneRow(
+    simdjson::padded_string_view input,
+    exec::VectorWriter<Any>& writer) {
+  SIMDJSON_ASSIGN_OR_RAISE(auto doc, simdjsonParse(input));
+  if (doc.is_null()) {
+    writer.commitNull();
+  } else {
+    SIMDJSON_TRY(ParseJsonTypedImpl<simdjson::ondemand::document&>::apply<kind>(
+        doc, writer.current(), true));
+    writer.commit(true);
+  }
+  return simdjson::SUCCESS;
+}
+
+class FromJsonExpr : public SpecialForm {
+ public:
+  /// @param type The target type of the cast expression
+  /// @param expr The expression to gerenate input
+  /// @param trackCpuUsage Whether to track CPU usage
+  FromJsonExpr(TypePtr type, ExprPtr&& expr, bool trackCpuUsage)
+      : SpecialForm(
+            type,
+            std::vector<ExprPtr>({expr}),
+            FromJsonCallToSpecialForm::kFromJson,
+            false /* supportsFlatNoNullsFastPath */,
+            trackCpuUsage) {
+    if (!isSupportedType(type)) {
+      VELOX_UNSUPPORTED("Unsupported type {}.", type->toString());
+    }
+    simdjsonErrorsToExceptions(errors_);
+  }
+
+  void evalSpecialForm(
+      const SelectivityVector& rows,
+      EvalCtx& context,
+      VectorPtr& result) override {
+    VectorPtr input;
+    inputs_[0]->eval(rows, context, input);
+    auto toType = std::const_pointer_cast<const Type>(type_);
+    apply(rows, input, context, toType, result);
+    // Return 'input' back to the vector pool in 'context' so it can be reused.
+    context.releaseVector(input);
+  }
+
+ private:
+  void computePropagatesNulls() override {
+    propagatesNulls_ = false;
+  }
+
+  // Peal data.
+  void apply(
+      const SelectivityVector& rows,
+      const VectorPtr& input,
+      exec::EvalCtx& context,
+      const TypePtr& toType,
+      VectorPtr& result) {
+    LocalSelectivityVector remainingRows(context, rows);
+
+    context.deselectErrors(*remainingRows);
+
+    LocalDecodedVector decoded(context, *input, *remainingRows);
+    auto* rawNulls = decoded->nulls(remainingRows.get());
+
+    if (rawNulls) {
+      remainingRows->deselectNulls(
+          rawNulls, remainingRows->begin(), remainingRows->end());
+    }
+
+    VectorPtr localResult;
+    if (!remainingRows->hasSelections()) {
```

Comment:
Can from_json be implemented as a vector function and registered as a special form like [DecimalRound](https://github.com/facebookincubator/velox/blob/main/velox/functions/sparksql/specialforms/DecimalRound.cpp)? What are the special considerations here to follow the implementation of 'cast'? Thanks.

Reply:
Before I want to to reuse the `JsonCastOperator`. But there are too many differences. Changed to vector fucntion.